front/lib/api/assistant/actions/retrieval.ts

```diff
@@ -805,3 +821,17 @@ export async function* runRetrieval(
     },
   };
 }
+
+async function datasourceLimitsReached(auth: Authenticator): Promise<boolean> {
+  const plan = auth.plan();
+  if (!plan) {
+    throw new Error("Unexpected missing plan");
+  }
+  if (!plan.limits.dataSources.count || plan.limits.dataSources.count < 0) {
+    return false;
+  }
+  const dataSourcesCount = (await getDataSources(auth)).filter(
```

Comment:
I am very worried that this can slow us down a lot. 
How many data sources are concerned by this limit? What triggered the need to do this task?

Reply:
No disrespect, but did you read the PR top comment? (with related task & opinion on why I think in that case it's fine?)

Reply:
You might be right, but it's better if you have read that before we discuss IMHO

Reply:
My initial feeling was that the count only happens when 1. there is a limit on the datasource count 2. there are static datasources concerned by the retrieval 

Presently, this would happen for 1. free users 2. that have custom assistants using a static datasource 3. only when they call those assistants 4. provided they haven't reached their message limit => seems rather limited

Reply:
TBH I am still weighing the tradeoff between enforcing the limits properly and shipping fast this low-value-add feature